crates/walrus-service/src/node/pending_metadata_cache.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! In-memory cache for metadata awaiting blob registration.
+
+use std::{
+    collections::{HashMap, hash_map::Entry},
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use tokio::sync::Mutex;
+use walrus_core::{BlobId, metadata::VerifiedBlobMetadataWithId};
+
+use super::metrics::NodeMetricSet;
+
+#[derive(Debug, Clone)]
+struct CachedMetadata {
+    inserted_at: Instant,
+    metadata: Arc<VerifiedBlobMetadataWithId>,
+}
+
+#[derive(Debug)]
+struct PendingMetadataCacheInner {
+    max_entries: usize,
+    ttl: Duration,
+    entries: HashMap<BlobId, CachedMetadata>,
+}
+
+impl PendingMetadataCacheInner {
+    fn new(max_entries: usize, ttl: Duration) -> Self {
+        Self {
+            max_entries,
+            ttl,
+            entries: HashMap::new(),
+        }
+    }
+
+    fn evict_expired(&mut self, now: Instant) {
+        if self.ttl.is_zero() {
+            return;
+        }
+        let ttl = self.ttl;
+        self.entries
+            .retain(|_, cached| now.saturating_duration_since(cached.inserted_at) < ttl);
+    }
+
+    fn get(&self, blob_id: &BlobId) -> Option<Arc<VerifiedBlobMetadataWithId>> {
+        self.entries
+            .get(blob_id)
+            .map(|cached| cached.metadata.clone())
+    }
+
+    /// Inserts metadata into the cache.
+    /// Returns `Err(())` when the cache is at capacity and the entry could not be stored after
+    /// evicting expired items.
+    fn insert(
+        &mut self,
+        blob_id: BlobId,
+        metadata: Arc<VerifiedBlobMetadataWithId>,
+    ) -> Result<(), ()> {
+        let now = Instant::now();
+        if self.max_entries == 0 {
+            return Err(());
+        }
+
+        if self.entries.len() >= self.max_entries {
+            self.evict_expired(now);
+            if self.entries.len() >= self.max_entries {
+                return Err(());
+            }
+        }
+
+        match self.entries.entry(blob_id) {
+            Entry::Occupied(mut entry) => {
+                entry.insert(CachedMetadata {
+                    inserted_at: now,
+                    metadata,
+                });
+                Ok(())
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(CachedMetadata {
+                    inserted_at: now,
+                    metadata,
+                });
+                Ok(())
+            }
+        }
```

Comment:
question: why does this read and modify? Seems can directly insert into the HashMap?